File: Python/DFS.py
class Node():
	def __init__(self, x):
		self.val = x
		self.childs = []

def DFS(node):
	stack = [node]
	marked = {}
	while len(stack) > 0:
		no  = stack.pop()
		if no not in marked:
			marked[no] = 1
			print(no.val)
            #stack += no.childs
			for p in no.childs:
				stack.append(p);


def main():
	a = Node(1)
	b = Node(2)
	c = Node(3)
	d = Node(4)
	e = Node(5)
	f = Node(6)
	g = Node(7)
	h = Node(8)
	i = Node(9)
	a.childs.append(b)
	a.childs.append(c)
	a.childs.append(d)
	b.childs.append(e)
	b.childs.append(f)
	c.childs.append(g)
	c.childs.append(i)
	d.childs.append(h)
	d.childs.append(b)
	e.childs.append(d)
	DFS(a)

if __name__ == '__main__':
	main()

File: Cpp/DFS.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <set>

using namespace std;

template<typename T>
class Node{
public:
	T val;
	vector<Node<T>*> childs;
	Node(T x){
		this->val = x;
	}
};


void DFS(Node<int> *node){
	stack<Node<int>*> s;
	s.push(node);
	set<Node<int>*> marked;
	while(s.empty() == false){
		Node<int> *no = s.top();
		s.pop();
		if (marked.find(no) == marked.end()){
			marked.insert(no);
			cout << no->val << " ";
			for(int i=0; i < no->childs.size(); i++){
				s.push(no->childs[i]);
			}
		}

	}
}

int main(){
	Node<int> a(1);
	Node<int> b(2);
	Node<int> c(3);
	Node<int> d(4);
	Node<int> e(5);
	Node<int> f(6);
	Node<int> g(7);
	Node<int> h(8);
	Node<int> i(9);
	a.childs.push_back(&b);
	a.childs.push_back(&c);
	a.childs.push_back(&d);
	b.childs.push_back(&e);
	b.childs.push_back(&f);
	c.childs.push_back(&g);
	c.childs.push_back(&i);
	d.childs.push_back(&h);
	d.childs.push_back(&b);
	e.childs.push_back(&d);
	DFS(&a);
	return 0;
}
